torchtitan/models/deepseek_v3/model/state_dict_adapter.py

```diff
@@ -52,6 +60,278 @@ def __init__(self, model_args: DeepSeekV3ModelArgs, hf_assets_path: str | None):
+        num_experts = grouped_expert_weight.shape[0]
+        start_index, end_index = self._caculate_indices_from_placements(
+            dim=0,
+            dim_size=num_experts,
+            dtensor_placements=dtensor_placements,
+            device_mesh=device_mesh,
+        )
+        assert (
+            start_index is not None and end_index is not None
+        ), "Start index and end index can not be None on dim-0!"
+
+        # Step 2: Store indices for potential future use in from_hf()
+        self.local_experts_indices[titan_abstract_key] = (start_index, end_index)
+
+        # Step 3: Create new placements for individual expert weights
+        new_placements = []
+        for i, name in enumerate(device_mesh.mesh_dim_names):
+            placement = dtensor_placements[i]
+            if placement.dim == 0:
+                # Convert dim-0 sharding to replication for individual experts
+                new_placements.append(Replicate())
+            elif isinstance(placement, Shard):
+                # Keep other shard dimensions (individual expert weight has 2D)
+                new_placements.append(Shard(placement.dim))
+            elif isinstance(placement, _StridedShard):
+                # Keep strided shard with same parameters
+                new_placements.append(
+                    _StridedShard(placement.dim, placement.split_factor)
+                )
+            else:
+                raise ValueError(f"Unsupported placement type: {type(placement)}")
+
+        # Step 4: Create individual expert DTensors
+        assert isinstance(
+            grouped_expert_weight, DTensor
+        ), "Expected DTensor for grouped expert weight"
+
+        local_grouped_weights = grouped_expert_weight._local_tensor
+        expected_local_experts = end_index - start_index
+
+        if local_grouped_weights.shape[0] != expected_local_experts:
+            raise ValueError(
+                f"Local tensor shape mismatch: expected {expected_local_experts} experts, "
+                f"got {local_grouped_weights.shape[0]}"
+            )
+
+        local_expert_tensors = {}
+        for expert_id in range(start_index, end_index):
+            expert_key = abstract_key.format(layer_id, expert_id)
+            local_expert_index = expert_id - start_index
+
+            # Extract individual expert weight and add batch dimension temporarily
+            expert_weight = local_grouped_weights[local_expert_index, :, :].unsqueeze(0)
+
+            # Create DTensor and remove batch dimension (experts dimension is removed)
+            expert_dtensor = DTensor.from_local(
+                expert_weight, device_mesh, new_placements, run_check=False
+            ).squeeze(0)
+
+            local_expert_tensors[expert_key] = expert_dtensor
+
+        return local_expert_tensors
+
+    def _concatenate_local_expert_weights(
+        self,
+        expert_weights_by_layer: dict[str, Any],
```

Comment:
We can also add description of this mapping, {layer: {abstract_key: {expert_id: tensor}}}, to the docstring.
```suggestion
        expert_weights_by_layer: dict[str, dict[str, dict[int, torch.Tensor]]],
```